llvm/test/Transforms/SimplifyCFG/X86/fake-use-considered-when-sinking.ll

```diff
@@ -0,0 +1,67 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt -p="simplifycfg<sink-common-insts>" -S < %s | FileCheck %s
+
+;; Verify that fake uses are not ignored when sinking instructions in
+;; SimplifyCFG; when a fake use appears in only one incoming block they prevent
+;; further sinking, and when identical fake uses appear on both sides they
+;; are sunk normally.
```

Comment:
Seems a bit fuzzy to me whether fake.use should block this optimisation or ignore the intrinsics and sink them into the suc. I don't find the [docs](https://llvm.org/docs/LangRef.html#id1872) particularly clarifying for this specific case.

You've been close to this work and have upstreamed it, so this is less me contesting the correct behaviour, more hoping for a bit more of rationale/explanation?

Reply:
The basic reason here is that fake uses must be treated as any other instruction by default. As an optimization we can create specific exceptions for fake uses in cases where we know it's either safe or beneficial to treat them differently, but we wouldn't want to do so "generally" as we do by skipping them in `getPrevNonDebugInstruction`.

For sinking common instructions specifically, I imagine it's safe to sink fake uses separate to the other instructions in their block, but that specialization should exist in the sinking code (and would be beyond the scope of this patch).